Implements delete project endpoint

p0lyn0mial opened the pull request:
**What this PR does / why we need it**: Implements delete a project HTTP endpoint

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1092 

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

p0lyn0mial committed:
sets up the owner of the project

makes sure that ProjecGroup.Name and ProjectGroup.Group with
proper values is attached to User object for the owner of the project.

p0lyn0mial committed:
creates RBAC Roles/Bindings for project's owners group

p0lyn0mial committed:
dep update

p0lyn0mial committed:
fix linter

p0lyn0mial committed:
implements HTTP Delete project endpoint

p0lyn0mial committed:
fix linter

p0lyn0mial committed:
add UT

p0lyn0mial commented on api/pkg/handler/project.go:
```diff
@@ -29,14 +30,11 @@ func createProjectEndpoint(projectProvider provider.ProjectProvider) endpoint.En
-				return nil, errors.NewAlreadyExists("Project", projectRq.Name)
-			}
-			return nil, err
+			return nil, errors.KubernetesErrorToHTTPError(err)
```
decided to slightly change the way we return errors.
since all errors returned from kubernetes API have well defined structure whey not to take advantage of it ?

mrIncompetent replied:
Dope idea. Something we should implement in every handler which might receive a k8s error!
Definitely something for a Friday-Cleanup

mrIncompetent replied:
Though i just realized that this should probably be handled in the responsible provider?
I dislike having the handler knowing about the kubernetes errors. Though just personal preference.

mrIncompetent commented on api/pkg/handler/project.go:
```diff
@@ -49,9 +47,19 @@ func getProjectsEndpoint() endpoint.Endpoint {
+		if !ok {
+			return nil, errors.NewBadRequest("invalid request")
+		}
+		if len(projectName) == 0 {
```
Isn't this check enough?
If the cast fails, the string will be empty

p0lyn0mial replied:
I think it is :)
If the cast fails, then we will return `BadRequest` line `54`

mrIncompetent replied:
Question is if its already enough checking for the empty string

mrIncompetent commented on api/pkg/provider/kubernetes/project.go:
```diff
@@ -51,7 +57,7 @@ func (p *ProjectProvider) New(user *kubermaticapiv1.User, projectName string) (*
 		for _, owner := range owners {
 			if owner.UID == user.UID && project.Spec.Name == projectName {
-				return nil, ErrAlreadyExist
+				return nil, kerrors.NewAlreadyExists(schema.GroupResource{Group: project.GroupVersionKind().Group, Resource: kubermaticapiv1.ProjectPlural}, projectName)
```
I would rather not pipe through the k8s errors here.
The provider should abstract the k8s specifics away

p0lyn0mial replied:
That was my initial idea but then I realised that we call all the providers `kubernetes` providers. They live under `provider/kubernetes/`. That's why they return k8s like errors :)

Kubernetes-like errors map very well to HTTP errors so in practice it means that most of the time they mach exactly to what we want. Plus we don't have to replicate the errors.

p0lyn0mial committed:
review changes

p0lyn0mial commented:
@mrIncompetent comments addressed. PTAL

I removed checking if `len(projectName) == 0` and moved `KubernetesErrorsToHTTPErrors` method to be internal to `delete` handler.

p0lyn0mial committed:
Merge branch 'master' into implement_delete_project_endpoint

p0lyn0mial committed:
Merge branch 'master' into implement_delete_project_endpoint

p0lyn0mial committed:
Merge branch 'master' into implement_delete_project_endpoint
